src/Symfony/Component/Serializer/Normalizer/DataUriNormalizer.php

```diff
@@ -0,0 +1,157 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Serializer\Normalizer;
+
+use Symfony\Component\HttpFoundation\File\File;
+use Symfony\Component\HttpFoundation\File\MimeType\MimeTypeGuesser;
+use Symfony\Component\HttpFoundation\File\MimeType\MimeTypeGuesserInterface;
+use Symfony\Component\Serializer\Exception\InvalidArgumentException;
+use Symfony\Component\Serializer\Exception\UnexpectedValueException;
+
+/**
+ * Normalizes an {@see \SplFileInfo} object to a data URI.
+ * Denormalizes a data URI to a {@see \SplFileObject} object.
+ *
+ * @author Kévin Dunglas <dunglas@gmail.com>
+ */
+class DataUriNormalizer implements NormalizerInterface, DenormalizerInterface
+{
+    /**
+     * @var MimeTypeGuesserInterface
+     */
+    private $mimeTypeGuesser;
+
+    public function __construct(MimeTypeGuesserInterface $mimeTypeGuesser = null)
+    {
+        if (null === $mimeTypeGuesser && class_exists('Symfony\Component\HttpFoundation\File\MimeType\MimeTypeGuesser')) {
+            $mimeTypeGuesser = MimeTypeGuesser::getInstance();
+        }
+
+        $this->mimeTypeGuesser = $mimeTypeGuesser;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function normalize($object, $format = null, array $context = array())
+    {
+        if (!$object instanceof \SplFileInfo) {
+            throw new InvalidArgumentException('The object must be an instance of "\SplFileInfo".');
+        }
+
+        $mimeType = $this->getMimeType($object);
+        $splFileObject = $this->extractSplFileObject($object);
+
+        $data = '';
+
+        $splFileObject->rewind();
+        while (!$splFileObject->eof()) {
+            $data .= $splFileObject->fgets();
+        }
+
+        if ('text' === explode('/', $mimeType, 2)[0]) {
+            return sprintf('data:%s,%s', $mimeType, rawurlencode($data));
+        }
+
+        return sprintf('data:%s;base64,%s', $mimeType, base64_encode($data));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function supportsNormalization($data, $format = null)
+    {
+        return $data instanceof \SplFileInfo;
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * Regex adapted from Brian Grinstead code.
+     *
+     * @see https://gist.github.com/bgrins/6194623
+     *
+     * @throws InvalidArgumentException
+     * @throws UnexpectedValueException
+     */
+    public function denormalize($data, $class, $format = null, array $context = array())
+    {
+        if (!preg_match('/^data:([a-z0-9]+\/[a-z0-9]+(;[a-z0-9\-]+\=[a-z0-9\-]+)?)?(;base64)?,[a-z0-9\!\$\&\\\'\,\(\)\*\+\,\;\=\-\.\_\~\:\@\/\?\%\s]*\s*$/i', $data)) {
```

Comment:
Uh-oh, i can't imagine how the /etc/shadow file could get exposed since the
prefix is predetermined, can you please give me some documentation?

On 20:14, Thu, Feb 4, 2016 Kévin Dunglas <notifications@github.com> wrote:

> In src/Symfony/Component/Serializer/Normalizer/DataUriNormalizer.php
> <https://github.com/symfony/symfony/pull/16164#discussion_r51913006>:
>
> > +        return $data instanceof \SplFileInfo;
> > +    }
> > +
> > +    /**
> > +     * {@inheritdoc}
> > +     *
> > +     * Regex adapted from Brian Grinstead code.
> > +     *
> > +     * @see https://gist.github.com/bgrins/6194623
> > +     *
> > +     * @throws InvalidArgumentException
> > +     * @throws UnexpectedValueException
> > +     */
> > +    public function denormalize($data, $class, $format = null, array $context = array())
> > +    {
> > +        if (!preg_match('/^data:([a-z0-9]+\/[a-z0-9]+(;[a-z0-9\-]+\=[a-z0-9\-]+)?)?(;base64)?,[a-z0-9\!\$\&\\\'\,\(\)\*\+\,\;\=\-\.\_\~\:\@\/\?\%\s]*\s*$/i', $data)) {
>
> There is two main reasons:
>
>    - this is very security sensitive (in my former implementation, it was
>    allowing to access to /etc/shadow for instance and, being not a security
>    expert, I prefer to be conservative - what about specials chars...)
>    - the PHP stream wrapper can generate an error if the data is invalid,
>    the regex prevent it
>
> However [image: :+1:] to make the regex as fast as possible
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/symfony/symfony/pull/16164/files#r51913006>.
>